src/sinks/amqp.rs

```diff
@@ -0,0 +1,615 @@
+        let sink = AmqpSink::new(self.clone(), cx.acker()).await?;
+        let hc = healthcheck(self.clone(), sink.channel.clone()).boxed();
+        Ok((super::VectorSink::Sink(Box::new(sink)), hc))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Any
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "amqp"
+    }
+}
+
+impl AmqpSink {
+    async fn new(config: AmqpSinkConfig, acker: Acker) -> crate::Result<Self> {
+        let (connection, channel) = config.connection.connect().await.map_err(|e| BuildError::AmqpCreateFailed { source: e })?;
+        Ok(AmqpSink {
+            _connection: connection,
+            channel: Arc::new(channel),
+            exchange: Template::try_from(config.exchange).context(ExchangeTemplate)?,
+            key_field: config.key_field,
+            encoding: config.encoding,
+            delivery_fut: FuturesUnordered::new(),
+            in_flight: FuturesUnordered::new(),
+            acker,
+            seq_head: 0,
+            seq_tail: 0,
+            pending_acks: HashSet::new(),
+        })
+    }
+
+    fn poll_delivery_fut(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        while !self.delivery_fut.is_empty() {
+            let result = Pin::new(&mut self.delivery_fut).poll_next(cx);
+            let (seqno, fut) = ready!(result).expect("`delivery_fut` is endless stream");
+            self.in_flight.push(Box::pin(async move {
+                let res = match fut.await {
+                    Ok(confirm) => {
+                        match confirm.await {
+                            Err(e) => Err(e),
+                            Ok(confirm) => {
+                                if let lapin::publisher_confirm::Confirmation::Nack(c) = confirm {
+                                    Ok(c)
+                                } else {
+                                    Ok(None)
+                                }
+                            }
+                        }
+                    }
+                    Err(error) => {
+                        Err(error)
+                    }
+                };
+                (seqno, Ok(res))
+            }));
+        }
+
+        Poll::Ready(())
+    }
+}
+
+impl Sink<Event> for AmqpSink {
+    type Error = ();
+
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match self.poll_delivery_fut(cx) {
+            Poll::Pending if self.delivery_fut.len() >= SEND_RESULT_LIMIT => Poll::Pending,
+            _ => Poll::Ready(Ok(())),
+        }
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: Event) -> Result<(), Self::Error> {
+        assert!(
+            self.delivery_fut.len() < SEND_RESULT_LIMIT,
+            "Expected `poll_ready` to be called first."
+        );
+
+        let exchange = self.exchange.render_string(&item).map_err(|missing_keys| {
+            error!(message = "Missing keys for exchange.", missing_keys = ?missing_keys);
```

Comment:
We should emit `TemplateRenderingFailed` event for this. See example:

https://github.com/timberio/vector/blob/b49d4339f8964072a527bf7cd82c97c63531010c/src/sinks/kafka.rs#L281-L287